File: 2015 Specimen Paper 2 - Ruby/PreRelease.rb
names = []
test1 = []
test2 = []
test3 = []
total = []
maxScores = [20,25,35]
score = 0
#Initializes the variables
30.times do
	print "Name:"
	names.append($stdin.gets.strip) #inputs the name of the student
	studentScore = []
	3.times do |num|
		begin
			print "Test #{num + 1} score:" #Input the test scores
			score = $stdin.gets.strip.to_i
		end while 0..maxScores[num] === score
		studentScore.append[score]
	end
	test1.append[studentScore[0]] #assign the scores to the arrays
	test2.append[studentScore[1]] #appends to the end
	test3.append[studentScore[2]]
	total.append(studentScore.inject(0,:+)) #appends the sum of the array
end
puts "Name | Test 1 | Test 2 | Test 3 | Total"
30.times {|num| puts "#{names[num]} | #{test1[num]} | #{test2[num]} | #{test3[num]} | #{total[num]}"}
#prints each student's name, scores and total
puts "Average = #{(total.inject(0,:+)) / 30}"
#average = sum/30
puts "Highest Scoring Student | Test 1 | Test 2 | Test 3 | Total"
mIndex = total.index((total.sort).last) #finds the index of the student with the highest total
puts "#{names[mIndex]} | #{test1[mIndex]} | #{test2[mIndex]} | #{test3[mIndex]} | #{total[mIndex]}"

File: 2015 Specimen Paper 2 - Python/PreRelease.py
names = []
test1 = []
test2 = []
test3 = []
total = []
totalMarks = 0
#initializes the arrays and variables

for num in range(30):
  names.append(input("Name: ")) #Appends the name of the student to the array
  score = int(input("Test 1 Score:")) #Inputs a value for test1
  while score > 20 or score < 0: #Validation
    print("Score is out of range") #Outputs error message
    score = int(input("Test 1 Score:")) #Inputs again
  else:
    test1.append(score) #Appends a validated score to the array
  score = int(input("Test 2 Score:")) #Same as above
  while score > 25 or score < 0:
    print("Score is out of range")
    score = int(input("Test 2 Score:"))
  else:
    test2.append(score)
  score = int(input("Test 3 Score:"))#Same as above
  while score > 35 or score < 0:
    print("Score is out of range")
    score = int(input("Test 3 Score:"))
  else:
    test3.append(score)
  total.append(test1[num] + test2[num] + test3[num])
  #Appends the total mark to the array
  totalMarks += total[num] #Adds student total to totalMarks

print("Name | Test 1 | Test 2 | Test 3") #Print column headers
for num in range(30):
  print(names[num],test1[num],test2[num],test3[num],total[num]) #Print each student's name, scores and total

print("Average")
print(totalMarks / 30) #Prints the class average

sortedTotal = sorted(total) #Duplicates and sorts the total score array
maxIndex = total.index(sortedTotal[29])
#Index of the student would be last in the sortedTotal array (29) and finds that student in the unsorted array
#This returns the position in the arrays the student is in

print("Highest Mark Student")
print("Name | Test 1 | Test 2 | Test 3") #Prints column headers
print(names[maxIndex],test1[maxIndex],test2[maxIndex],test3[maxIndex]) #Prints out student name, scores and total